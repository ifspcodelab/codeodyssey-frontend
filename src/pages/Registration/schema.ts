import i18n from "../../locales/i18n";
import * as yup from "yup"

export const schema = yup.object({
    name: yup.string().required().min(5).max(100),
    email: yup.string().required().email().max(350),
    password: yup.string().required().min(8).max(64)
        .matches(/\d+/, i18n.t('registration.form.validation.password.number'))
        .matches(/[a-z]+/, i18n.t('registration.form.validation.password.lowercase'))
        .matches(/[A-Z]+/, i18n.t('registration.form.validation.password.uppercase'))
        .matches(/[\W_]+/, i18n.t('registration.form.validation.password.special')),
    terms: yup.boolean().oneOf([true], i18n.t('registration.form.validation.termsCheckbox'))
}).required()