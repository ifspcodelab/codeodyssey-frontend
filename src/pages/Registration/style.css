.formContainer {
    width: 400px;
    margin: 20px auto;
    height: 200px;
}

form input {
    width: 100%;
    padding: 0;
}

.formInput * {
    display: block;
}

.formInput p {
    color: #FF0000;
}

.formInput input {
    padding-left: 4px;
}

.checkbox {
    display: block;
    align-items: center;
    margin: 10px auto;
    width: 100%;
}

.checkbox input {
    margin-right: 10px;
    width: 15px;
    aspect-ratio: 1;
}

.checkbox p {
    color: #FF0000;
}

#terms {
    position: relative;
    top: -2px;
}

#buttons {
    display: flex;
    justify-content: space-between;
}

#footer {
    display: flex;
    left: 50%;
    background-color: aqua;
}