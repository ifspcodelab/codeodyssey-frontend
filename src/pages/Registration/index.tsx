import PageHeader from "../../components/PageHeader";
import PageFooter from "../../components/PageFooter";
import './style.css'
import {useNavigate} from "react-router-dom"
import {useForm} from "react-hook-form";
import {yupResolver} from "@hookform/resolvers/yup";
import {Trans, useTranslation} from "react-i18next";
import {CreateUserResponse} from "../../core/models/CreateUserResponse";
import {schema} from "./schema";
import {useApi} from "../../core/hooks/useApi";
import {
    Button,
    Checkbox,
    Container,
    FormControlLabel,
    FormGroup,
    Grid,
    TextField,
    Link,
    Typography
} from "@mui/material";


function Registration() {
    const {t} = useTranslation()
    const { register, handleSubmit, formState: { errors } } = useForm({ resolver: yupResolver(schema)})
    const navigate = useNavigate()
    const api = useApi()

    const onSubmit = async (data: CreateUserResponse) => {
        const response = await api.register(data.name, data.email, data.password)
        if (response) {
            navigate('/resend-email', { state: { data: data.email }})
        }
    }

    return (
        <>
            <Container maxWidth="md">
                <div className="formContainer">
                    <PageHeader title={t('registration.title')} text={t('registration.text')}/>
                    <form onSubmit={handleSubmit(onSubmit)}>
                        <Grid container spacing={1} rowSpacing={2}>
                            <Grid item xs={12}>
                                <TextField
                                    sx={{ width: "100%" }}
                                    {...register("name")}
                                    label={t("registration.form.name")}
                                    variant="outlined"
                                    error={!!errors.name}
                                    helperText={errors.name && <span>{errors.name.message}</span> }
                                    inputProps={{ "data-testid": "nameField" }}
                                    aria-labelledby="name"
                                />
                            </Grid>
                            <Grid item xs={12}>
                                <TextField
                                    sx={{ width: "100%" }}
                                    {...register("email")}
                                    label={t("registration.form.email")}
                                    variant="outlined"
                                    error={!!errors.email}
                                    helperText={errors.email && <span>{errors.email.message}</span> }
                                    inputProps={{ "data-testid": "emailField" }}
                                    aria-labelledby="email"
                                />
                            </Grid>
                            <Grid item xs={12}>
                                <TextField
                                    sx={{ width: "100%" }}
                                    {...register("password")}
                                    label={t("registration.form.password")}
                                    variant="outlined"
                                    error={!!errors.password}
                                    helperText={errors.password && <span>{errors.password.message}</span> }
                                    inputProps={{ "data-testid": "passwordField" }}
                                    aria-labelledby="password"
                                    type="password"
                                />
                            </Grid>
                            <Typography id="emailMessage">* {t("registration.form.email_message")}</Typography>
                            <Grid item xs={12}>
                                <FormGroup aria-labelledby="terms" >
                                    <FormControlLabel {...register("terms")} control={<Checkbox  />} label={
                                        <Trans i18nKey="registration.form.termsCheckbox">
                                            I have read and agree with the
                                            <Link href="/terms-of-use" target="_blank" rel="noopener noreferrer">Terms of Use</Link>
                                            and
                                            <Link href="/privacy-policy" target="_blank" rel="noopener noreferrer">Privacy Policy</Link>
                                        </Trans>
                                    } />
                                    <Typography color="error">{errors.terms?.message}</Typography>
                                </FormGroup>
                            </Grid>
                            <Grid item xs={12}>
                                <div id="registration-menu">
                                    <Button data-testid="registerButton" type="submit" variant="contained" size="large">{t('registration.form.submit')}</Button>
                                    <Link data-testid="loginLink" href="/login" rel="noopener noreferrer" underline="hover">
                                        {t('registration.form.login')}
                                    </Link>
                                </div>
                            </Grid>
                        </Grid>
                    </form>
                    <PageFooter text={t('registration.footer')}/>
                </div>
            </Container>
        </>
    );
}

export default Registration;