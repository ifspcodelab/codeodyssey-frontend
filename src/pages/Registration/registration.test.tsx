import "@testing-library/jest-dom";
import { render, fireEvent } from "@testing-library/react";
import {describe, test, vitest} from "vitest";
import Registration from "./index";
import {BrowserRouter} from "react-router-dom";
import {schema} from "../Registration/index.tsx"

const submitButtonId = "submitButton"

describe("Registration", () => {

   test("Should be able to see the Page Header title on the screen", () => {
       const { getByText } = render(
           <BrowserRouter>
               <Registration/>
           </BrowserRouter>
       );

       expect(getByText("Registration")).toBeInTheDocument();
   })

    test("Should be able to see the submit button text", () => {
        const { getByTestId } = render(
            <BrowserRouter>
                <Registration/>
            </BrowserRouter>
        );

        expect(getByTestId(submitButtonId)).toBeInTheDocument();
    })

    test("requires user's name", async () => {
        await expect(schema.validateAt('name', {})).rejects.toMatch(/name is a required field/)
        await expect(schema.validateAt('name', {name: "inv"})).rejects.toMatch(/name must be at least 5 characters/)
        await expect(schema.validateAt('name', {name: "valid name"})).resolves.toBeTruthy()
    })

    test("Should be able to send registration request", () => {
        const { getByTestId } = render(
            <BrowserRouter>
                <Registration/>
            </BrowserRouter>
        );

        fireEvent.click(getByTestId(submitButtonId));
    })

    test("Should send request with data after form submission", async () => {
        const { getByLabelText, getByTestId } = render(
            <BrowserRouter>
                <Registration />
            </BrowserRouter>
        );

        const inputName = getByLabelText('Name');
        const inputEmail = getByLabelText('Email');
        const inputPassword = getByLabelText('Password');
        const inputTerms = getByLabelText('I have read and agree with the Terms of Use and Privacy Policy')
        const submitButton =getByTestId('submitButton');

        fireEvent.change(inputName, { target: { value: 'John Doe' } });
        fireEvent.change(inputEmail, { target: { value: 'johndoe@email.com' } });
        fireEvent.change(inputPassword, { target: { value: 'Password@01' } });
        fireEvent.change(inputTerms, { target: {value: true}})

        fireEvent.click(submitButton);

        const result = await vitest.fn();

        expect(result).toMatchSnapshot();
    })

    test("Should send to login page after clicking the login button", async () => {
        const { getByText, getByTestId } = render(
            <BrowserRouter>
                <Registration />
            </BrowserRouter>
        );

        const loginButton = getByTestId("loginButton")

        fireEvent.click(loginButton)

        const result = await vitest.fn();

        expect(result).toMatchSnapshot("Login Page");
    })
});