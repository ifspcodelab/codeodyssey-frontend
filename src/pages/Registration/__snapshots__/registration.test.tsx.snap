// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Registration > Should send request with data after form submission 1`] = `[MockFunction spy]`;

exports[`Registration > Should send to login page after clicking the login button > Login Page 1`] = `[MockFunction spy]`;
